src/serac/numerics/functional/tuple_tensor_dual_functions.hpp

```diff
@@ -665,4 +676,39 @@ auto solve_scalar_equation(function&& f, double x0, double tolerance, double low
   }
 }
 
+/**
+ * @brief Finds a root of a vector-valued nonlinear function
+ *
+ * Uses Newton-Raphson iteration.
+ *
+ * @tparam function Type for the functor object
+ * @tparam n Vector dimension of the equation
+ * @param f A callable representing the function of which a root is sought. Must take an n-vector
+ * argument and return an n-vector
+ * @param x0 Initial guess for root. Must be an n-vector.
+ * @return A root of @p f.
+ */
+template <typename function, int n>
+auto find_root(function&& f, tensor<double, n> x0)
```

Comment:
we have these two methods `find_root` and `solve_scalar_equation` that are equivalent mathematically, I wonder if they might be combined in some way. 

One possible way is to overload on the type of the options:
```cpp
struct NewtonOptions {
  double tolerance;
  int max_iterations;
};

template <typename function, int n>
auto find_root(function&& f, tensor<double, n> x0, NewtonOptions opt) {
  ...
}

struct NewtSafeOptions {
  double tolerance;
  int max_iterations;
  double bounds[2];
};

template <typename function, int n>
auto find_root(function&& f, double x0, NewtSafeOptions opt) {
  ...
}
```
I don't think this needs to happen in this PR, but it might be worthwhile in the longer run

Reply:
I thought about this, too, and agree we should spend some time thinking about how much should be unified. One could argue that the top-level solvers for the finite element systems are also equivalent. To me, the scalar function solver is the most different. The bounds are not really options, for example (and would be meaningless for n > 1). But your point is well taken.

Reply:
just ran a couple tests on 2 machines:
- `matrix_mult_scalar` implements operator* by doing `B(i,j) = A(i,j) * scale);`
- `matrix_div1_scalar` implements operator/ by doing `B(i,j) = A(i,j) / scale);`
- `matrix_div2_scalar` implements operator/ by doing `inv_scale = 1.0 / scale;` first, then `B(i,j) = A(i,j) * inv_scale);`

Desktop x86 processor (i9-13900KF, -O3)
```
-------------------------------------------------------------
Benchmark                   Time             CPU   Iterations
-------------------------------------------------------------
matrix_mult_scalar      0.527 ns        0.527 ns   1000000000
matrix_div1_scalar       3.53 ns         3.53 ns    198139978
matrix_div2_scalar      0.529 ns        0.529 ns   1000000000
```

Desktop x86 processor (i9-13900KF, -O3 -ffast-math -march=native)
```
-------------------------------------------------------------
Benchmark                   Time             CPU   Iterations
-------------------------------------------------------------
matrix_mult_scalar      0.356 ns        0.356 ns   1000000000
matrix_div1_scalar      0.365 ns        0.365 ns   1000000000
matrix_div2_scalar      0.363 ns        0.363 ns   1000000000
```

Apple arm processor (M1, -O3)
```
-------------------------------------------------------------
Benchmark                   Time             CPU   Iterations
-------------------------------------------------------------
matrix_mult_scalar      0.786 ns        0.786 ns    701262272
matrix_div1_scalar      0.788 ns        0.788 ns    890200168
matrix_div2_scalar      0.786 ns        0.786 ns    890381338
```

Apple arm processor (M1, -O3  -ffast-math)
```
-------------------------------------------------------------
Benchmark                   Time             CPU   Iterations
-------------------------------------------------------------
matrix_mult_scalar      0.320 ns        0.320 ns   1000000000
matrix_div1_scalar      0.313 ns        0.313 ns   1000000000
matrix_div2_scalar      0.314 ns        0.314 ns   1000000000
```

----------

so, empirically: 
- no difference in either case when `-ffast-math` enabled
- the x86 version without `-ffast-math` is considerably slower
- all implementations on the arm processor are identical 

Either way, we should probably just implement the `operator/` overloads to compute `1.0 / scale` first, since we can't assume `-ffast-math` is always enabled.

Reply:
Awesome! Thanks for this valuable data point. Agree with your conclusion.

In the LLVM [fast-math details](https://llvm.org/docs/LangRef.html#fast-math-flags), note the `arcp` sub-option, which explicitly allows converting division to multiplication by the reciprocal. I think this sub-option, as well as all the others, are activated when clang is executed with `-ffast-math`, which would explain the identical results on x86 in that mode.

Some of the sub-options tradeoffs are generally acceptable for us, others in restricted cases, but I think the `reassoc` one will ruin the numerical stability of many algorithms (in the floating-point arithmetic sense). If it's true that all these options are activated in one fell swoop, we would need to use `-ffast-math` *very* narrowly.